### ISSUE TITLE
Automatically determine dependencies to guarantee reproducible builds

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#957]
Currently, the developer has to manually list dependencies and transitive dependencies in package.json.  Here is an example from states of matter:

```json
  "phet": {
    "colorProfile": true,
    "requirejsNamespace": "STATES_OF_MATTER",
    "phetLibs": [
      "nitroglycerin"
    ],
```

Note the `phetLibs: "nitroglycerin"`.  If this was omitted, the build would still work but we would be missing important SHAs and could no longer guarantee reproducible builds.  For instance, try removing that line and run `grunt --brands=phet` and notice the simulation builds but with an incomplete dependencies.json file.

@chrisklus @zepumph and I identified this problem while considering https://github.com/phetsims/studio/issues/162, in which we can now use module dependencies for Studio (and hope to use them in other wrappers in the future).  Once this issue is complete, we will need to get this working for wrappers.  For instance, we could output studio/dependencies.json.  Another way would be to combine all wrappers dependencies.json into the main dependencies.json.

**Comment by zepumph** [comment]
@pixelzoom ran into this only once FEL was being built by the build server for RC. I believe we should add an assertion more eagerly than the work here to auto detect.

**Comment by pixelzoom** [comment]
I also ran into this during Gas Properties RC build last week.

**Comment by zepumph** [comment]
Hmmm. Looks like this is a duplicate of https://github.com/phetsims/chipper/issues/983, and that we actually have this assertion in place, it just broke when we moved to using typescript, and thus modules are now found in `chipper/dist/js`. 

I'll update this block here: 

https://github.com/phetsims/chipper/blob/d9c7d31bfe6ce53879a9d5c21fc4f12d50b895b6/js/grunt/buildRunnable.js#L136-L143

**Comment by zepumph** [comment]
Updated in this commit. I tested with this command to check that all is up to date currently on main.

`for-each.sh perennial/data/active-runnables grunt --noTSC --lint=false --minify.minify=false --brands=phet > sandbox/output.txt`

This found issues in MSAB and PSD. @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
To test, I removed the "bamboo" dependency from gas-properties/package.json file, then attempted a `grunt build`.   I expected the build to fail, and indeed it did:

>  AssertionError [ERR_ASSERTION]: repo bamboo missing from package.json's phetLibs for chipper/dist/js/bamboo/js/ChartTransform.js

So this is behaving as expected, and catching this problem much earlier in the process. Thanks! Closing.
